source/develop/release-management/features/virt/tpm-device.md

```diff
@@ -0,0 +1,154 @@
+- Let swtpm store the data directly to the shared storage.
+- Some combination of the options above.
+
+Different options are discussed in the following subsections.  We can probably assume that TPM data updates are not going to be very frequent.  We mustn't forget about reboots, snapshots, imports and exports.
+
+#### Using API calls
+
+TPM data can be stored in the Engine database and transferred in both the directions using Vdsm API calls.  It can be provided to the VM as an additional parameter to VM.create call.  Data changes can be reported back to Engine using:
+
+- Events sent from Vdsm.
+- As a return value parameter of VM.destroy call.
+- As an additional value in VM stats.
+- On a request from Engine using a newly introduced Vdsm API call.
+
+TPM data can be copied to swtpm data location before the VM is created.  Once VM is started, updated TPM data can be read from the swtpm data location.  A problem is how to read data from swtpm consistently, avoiding collisions with swtpm writes.  Considering TPM data is small, written locally and not updated frequently, a good enough approach should be to have a periodic Vdsm job watching for swtpm data changes.  If it detects a data change, it keeps watching the data and once it detects the same data in two consecutive job runs, it can report it as changed data relatively safely (unless there is some fatal condition such as a full local disk).  This mechanism would be common to all the approaches except for shared storage directly used by swtpm.
+
+There are some problems specific to this approach though:
+
+- Events are not reliable and Vdsm can't be sure Engine has received and processed the updated data.
+- The same applies to reporting the data in VM stats, unless we are willing to send the data, even when unmodified, on each VM stats query.
+- Data read after the VM stops can be considered stable and reported directly.  However, if it is attached to VM.destroy response and the response fails for any reason, it can't be reported again in a response to a repeated VM.destroy call, because the data is already deleted after the libvirt domain has been undefined.
+- A specific API call would solve most of the problems.  Vdsm could include TPM data into the computation of the VM device hash sent within VM stats and Engine could request TPM data itself if the device hash changes.  Alternatively, Engine could make the API call periodically, providing TPM data hashes as an argument, and Vdsm would respond only with data not matching the hashes.  An additional call would happen in either case just before calling VM.destroy to get the ultimate final version of TPM data.
+- It is fully dependent on Engine cooperation, it's not possible to start a VM without Engine explicitly providing the data.  But if we needed just to restart a VM on the same host without Engine then we could retain TPM data by either not undefining the domain XML or by making a local copy of the data between the restarts.
```

Comment:
We cannot start a vm on same host without enigne since vm xml and metadata is deleted when stooping a vm.

I think the only way that will work is to have new API called by engine to fetch tpm data, and 
split destroying a vm to 2 steps - stop the vm and undefine the vm. This way the vm data is not
lost.